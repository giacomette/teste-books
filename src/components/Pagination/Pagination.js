import React from 'react';
import PropTypes from 'prop-types';
import { Pagination, PaginationLink, PaginationItem } from 'reactstrap';

const PaginationComponent = ({ currentPage, perPage, total, onChangePage }) => {
  const paginations = [];
  const totalPages = total ? Math.ceil(total / perPage) : 0;

  if (!totalPages) return null;

  for (let page = 1; page <= totalPages; page++) {
    paginations.push(page);
  }

  return (
    <Pagination aria-label="Page navigation">
      <PaginationItem disabled={totalPages === 1 || currentPage === 1}>
        <PaginationLink first onClick={() => onChangePage(1)} />
      </PaginationItem>
      <PaginationItem disabled={currentPage === 1}>
        <PaginationLink
          previous
          onClick={() => onChangePage(currentPage - 1)}
        />
      </PaginationItem>
      {paginations.map((page, key) => (
        <PaginationItem
          key={key}
          disabled={currentPage === page}
          className={currentPage === page ? 'active' : ''}
        >
          <PaginationLink onClick={() => onChangePage(page)}>
            {page}
          </PaginationLink>
        </PaginationItem>
      ))}
      <PaginationItem disabled={currentPage === totalPages}>
        <PaginationLink next onClick={() => onChangePage(currentPage + 1)} />
      </PaginationItem>
      <PaginationItem disabled={currentPage === totalPages}>
        <PaginationLink last onClick={() => onChangePage(totalPages)} />
      </PaginationItem>
    </Pagination>
  );
};

PaginationComponent.propTypes = {
  currentPage: PropTypes.number.isRequired,
  perPage: PropTypes.number.isRequired,
  total: PropTypes.number.isRequired,
  onChangePage: PropTypes.func.isRequired
};

export default PaginationComponent;
