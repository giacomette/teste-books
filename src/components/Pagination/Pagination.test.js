import React from 'react';
import Pagination from './Pagination';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

describe('<Pagination>', () => {
  it('deve haver alteração na estrutura do componente após um click', () => {
    let page = 1;

    const component = renderer.create(
      <Pagination
        currentPage={page}
        onChangePage={v => (page = v)}
        total={20}
        perPage={15}
      />
    );
    let tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('deve criar 7 itens na paginação', () => {
    let page = 1;
    let total = 45;

    const component = mount(
      <Pagination
        currentPage={page}
        onChangePage={v => (page = v)}
        total={total}
        perPage={15}
      />
    );
    const pages = component.find('.page-item');

    // 3 paginas + 4 botões direcionais
    expect(pages.length).toBe(7);
  });

  it('deve haver somento um item em active', async () => {
    let page = 1;
    let total = 45;

    const component = await mount(
      <Pagination
        currentPage={page}
        onChangePage={v => (page = v)}
        total={total}
        perPage={15}
      />
    );

    expect(component.find('.active.page-item').length).toBe(1);
  });

  it('deve atualizar as props', () => {
    let page = 1;
    let total = 45;

    let component = mount(
      <Pagination
        currentPage={page}
        onChangePage={v => (page = v)}
        total={total}
        perPage={15}
      />
    );

    expect(component.props().currentPage).toBe(1);

    component.setProps({ currentPage: 2 });

    expect(component.props().currentPage).toBe(2);
  });

  it('deve trocar a pagina ativa ao mudar a propriedade currentPage', () => {
    let page = 1;
    const total = 45;

    const component = mount(
      <Pagination
        currentPage={page}
        onChangePage={v => (page = v)}
        total={total}
        perPage={15}
      />
    );

    component.setProps({ currentPage: 2 });

    const active = component.find('.active.page-item').text();

    expect(active).toBe('2');
  });

  it('não deve trocar a pagina ativa ao atribuir a propriedade currentPage o mesmo valor atual', () => {
    let page = 1;
    let total = 45;

    let component = mount(
      <Pagination
        currentPage={page}
        onChangePage={v => (page = v)}
        total={total}
        perPage={15}
      />
    );

    component.setProps({ currentPage: 1 });

    let active = component.find('.active.page-item').text();

    expect(active).toBe('1');
  });

  it('Deve dimunuir o numero de paginas ao diminuir ao atualizar a propriedade de total', () => {
    let page = 1;
    const total = 45;

    const component = mount(
      <Pagination
        currentPage={page}
        onChangePage={v => (page = v)}
        total={total}
        perPage={15}
      />
    );

    component.setProps({ total: 25 });

    const pages = component.find('.page-item');

    // 2 paginas + 4 botões direcionais
    expect(pages.length).toBe(6);
  });
});
