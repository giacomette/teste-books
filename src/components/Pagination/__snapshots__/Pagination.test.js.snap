// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination> deve haver alteração na estrutura do componente após um click 1`] = `
<nav
  aria-label="Page navigation"
  className=""
>
  <ul
    className="pagination"
  >
    <li
      className="page-item disabled"
    >
      <button
        aria-label="First"
        className="page-link"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          «
        </span>
        <span
          className="sr-only"
        >
          First
        </span>
      </button>
    </li>
    <li
      className="page-item disabled"
    >
      <button
        aria-label="Previous"
        className="page-link"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          ‹
        </span>
        <span
          className="sr-only"
        >
          Previous
        </span>
      </button>
    </li>
    <li
      className="active page-item disabled"
    >
      <button
        className="page-link"
        onClick={[Function]}
      >
        1
      </button>
    </li>
    <li
      className="page-item"
    >
      <button
        className="page-link"
        onClick={[Function]}
      >
        2
      </button>
    </li>
    <li
      className="page-item"
    >
      <button
        aria-label="Next"
        className="page-link"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          ›
        </span>
        <span
          className="sr-only"
        >
          Next
        </span>
      </button>
    </li>
    <li
      className="page-item"
    >
      <button
        aria-label="Last"
        className="page-link"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          »
        </span>
        <span
          className="sr-only"
        >
          Last
        </span>
      </button>
    </li>
  </ul>
</nav>
`;
